# OILYGIANT. Finding the best location for a new oil well

## Contents  <a id='back'></a>

* [Introduction](#intro) 
* [1. Data exploration](#data_review)
* [2. Model evaluatio](#model_evaluation)
* [3. Profit calculation](#benefit_calcutions)
* [Conclusions](#end)

# Introduction <a id='intro'></a>

The mining company OilyGiant wants to find the best location for a new well by analyzing the parameters of oil wells in different regions (oil quality and volume of reserves).

## Objectives

* Validate an oil reserve volume prediction model for future OilyGiant operations.
* Calculate the profit for a set of selected oil wells.
* Calculate the risks and profits for each exploitation region.

[Back to Contents](#back)

## 1. Data Exploration <a id='data_review'></a>

## Download and prepopulate data

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
import numpy as np
from scipy import stats as st
 


In [2]:
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv')


In [3]:

geo_data_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [4]:
geo_data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [5]:

geo_data_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [6]:
geo_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [7]:

geo_data_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [8]:
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


On this occasion we can observe that all the data of the characteristics f0, f1 and f2, are of floating numerical type and no missing data are appreciated, so we can indicate that we can work with them.

[Back to Contents](#back)

## 2. Model evaluation <a id='model_evaluation'></a>

### TRAIN AND TEST THE MODEL FOR EACH REGION.

In this occasion we have as we mentioned we will work with each one of the dataframe, so the whole process of data division in training sets and validation set, model testing, RECM predictions of the model, we have to perform exactly the same with each one of the data frame geo_data_0, geo_data_1 and geo_data_2, so we propose to design a function that allows us to perform each one of the calculations and to obtain

In [9]:
def modeling_df(df_region):

# The average volume of reserves is calculated

    mean_volumen = df_region['product'].mean()
    

# The data is divided into a training set and a validation set in a ratio of 75:25.

    features = df_region.drop(columns =['product','id'], axis=1)
    target = df_region['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345)

# The model is chosen

    best_result = 10000
    best_est = 0
    best_depth = 0

    for est in range(10, 51, 10):
        for depth in range (1, 11):
            model_forest = RandomForestRegressor(random_state=12345 , n_estimators=est, max_depth=depth) 

# The model is trained and predictions are made for the validation set.

            model_forest.fit(features_train, target_train) 

            predictions_valid = model_forest.predict(features_valid) 

            result = mean_squared_error(target_valid, predictions_valid)**0.5

            if result < best_result:
                best_result = result
                best_est = est
                best_depth = depth

    
    result_1 = "El volúmen promedio de las reservas previstas son: " + str(mean_volumen)
    result_2 = "RECM del mejor modelo en el conjunto de validación es : " + str(best_result)
    result_3 = "best_depth:"+ str(depth)
    result_4 = "n_estimators:" + str(best_est)

    return result_1, result_2, result_3, result_4




In [10]:
modeling_df(geo_data_0)

('El volúmen promedio de las reservas previstas son: 92.50000000000001',
 'RECM del mejor modelo en el conjunto de validación es : 37.112242359808455',
 'best_depth:10',
 'n_estimators:50')

In [11]:
modeling_df(geo_data_1)

('El volúmen promedio de las reservas previstas son: 68.82500000000002',
 'RECM del mejor modelo en el conjunto de validación es : 0.7178000901597482',
 'best_depth:10',
 'n_estimators:30')

In [12]:
modeling_df(geo_data_2)

('El volúmen promedio de las reservas previstas son: 95.00000000000004',
 'RECM del mejor modelo en el conjunto de validación es : 37.70876829867544',
 'best_depth:10',
 'n_estimators:50')

Now the linear regression model is reviewed to determine the best model, this time the average volume of the regions and RECM of this model will be obtained.

In [13]:
def linear_df(df_region):

    mean_volumen = df_region['product'].mean()

# The data is divided into a training set and a validation set in a ratio of 75:25.

    features_df = df_region.drop(columns =['product','id'], axis=1)
    target_df = df_region['product']
    
    features_train, features_valid, target_train, target_valid = train_test_split(
        features_df, target_df, test_size=0.25, random_state=12345)
    
# Set the model to use
    model_linear = LinearRegression()  
# Train the model in the training set
    model_linear.fit(features_train, target_train) 
# predictions are obtained and added as column in the features df
    predictions_valid = model_linear.predict(features_valid) 
    
    features_valid['predictions']= predictions_valid

    target= target_valid
    predictions= features_valid['predictions']

    result = mean_squared_error(target_valid, predictions_valid)**0.5 

    print("El volúmen promedio de las reservas previstas son: " + str(mean_volumen)) 
    print("RECM del modelo de regresión lineal en el conjunto de validación:" + str(result))

    return target, predictions



In [14]:
linear_df(geo_data_0)

El volúmen promedio de las reservas previstas son: 92.50000000000001
RECM del modelo de regresión lineal en el conjunto de validación:37.5794217150813


/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


(71751     10.038645
 80493    114.551489
 2655     132.603635
 53233    169.072125
 91141    122.325180
             ...    
 12581    170.116726
 18456     93.632175
 73035    127.352259
 63834     99.782700
 43558    177.821022
 Name: product, Length: 25000, dtype: float64,
 71751     95.894952
 80493     77.572583
 2655      77.892640
 53233     90.175134
 91141     70.510088
             ...    
 12581    103.037104
 18456     85.403255
 73035     61.509833
 63834    118.180397
 43558    118.169392
 Name: predictions, Length: 25000, dtype: float64)

In [15]:
linear_df(geo_data_1)

El volúmen promedio de las reservas previstas son: 68.82500000000002
RECM del modelo de regresión lineal en el conjunto de validación:0.893099286775617


/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


(71751     80.859783
 80493     53.906522
 2655      30.132364
 53233     53.906522
 91141      0.000000
             ...    
 12581    137.945408
 18456    110.992147
 73035    137.945408
 63834     84.038886
 43558     53.906522
 Name: product, Length: 25000, dtype: float64,
 71751     82.663314
 80493     54.431786
 2655      29.748760
 53233     53.552133
 91141      1.243856
             ...    
 12581    136.869211
 18456    110.693465
 73035    137.879341
 63834     83.761966
 43558     53.958466
 Name: predictions, Length: 25000, dtype: float64)

In [16]:
linear_df(geo_data_2)

El volúmen promedio de las reservas previstas son: 95.00000000000004
RECM del modelo de regresión lineal en el conjunto de validación:40.02970873393434


/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


(71751     61.212375
 80493     41.850118
 2655      57.776581
 53233    100.053761
 91141    109.897122
             ...    
 12581     28.492402
 18456     21.431303
 73035    125.487229
 63834     99.422903
 43558    127.445075
 Name: product, Length: 25000, dtype: float64,
 71751     93.599633
 80493     75.105159
 2655      90.066809
 53233    105.162375
 91141    115.303310
             ...    
 12581     78.765887
 18456     95.603394
 73035     99.407281
 63834     77.779912
 43558    129.032417
 Name: predictions, Length: 25000, dtype: float64)

It can be observed that the root mean square error (RECM) is lower in the geo_data_1 dataframe and the average of this is the one with the lowest value, while the other two average volumes of the dataframe oscillate between 92 and 95, with a similar RECM between them. 

Observing the results of the linear regression model, similar results to those previously obtained are obtained, where in geo_data_1 the lowest RECM is obtained.

Considering this way, it can be observed that there is less error in the linear regression model in order to start the analysis of the product obtained in the wells of the different regions.

[Back to Contents](#back)

## 3. Profit calculation <a id='benefit_calcutions'></a>

### Profit calculation by region

We will calculate the volume to develop a new well without losses, total average profit for each of the regions, expenses per well and the average income from the volume obtained, in this case it is a general review of the data without the model to be able to appreciate more clearly what we are working with.

In [17]:
def metricas_beneficio(df_region):

# Se almacenan los valores clave en variables
    budget = 100000000
    well_per_budget = 200
    cost_per_well = budget / well_per_budget
    product_price = df_region['product']*4500

# Se calcula el volumen promedio de las reservas 
    mean_volumen = df_region['product'].mean()

# Se calcula el volumen para desarrollar un nuevo pozo sin pérdidas
    product_mean = product_price.mean()
    volumen_no_losses = budget /(well_per_budget*4500)
    
    profit = product_price - cost_per_well
    profit_mean= profit.mean()


    return "El volumen para desarrollar un nuevo pozo sin pérdidas: " + str(volumen_no_losses), "El volumen promedio de las reservas previstas son: " + str(mean_volumen), 'El gasto por pozo es de: '+str(cost_per_well), 'El ingreso promedio del volumen obtenido es de: '+str(product_mean),'El beneficio promedio es de: '+str (profit_mean)
    

In [18]:
metricas_beneficio(geo_data_0)

('El volumen para desarrollar un nuevo pozo sin pérdidas: 111.11111111111111',
 'El volumen promedio de las reservas previstas son: 92.50000000000001',
 'El gasto por pozo es de: 500000.0',
 'El ingreso promedio del volumen obtenido es de: 416250.00000000006',
 'El beneficio promedio es de: -83749.99999999994')

In [19]:
metricas_beneficio(geo_data_1)


('El volumen para desarrollar un nuevo pozo sin pérdidas: 111.11111111111111',
 'El volumen promedio de las reservas previstas son: 68.82500000000002',
 'El gasto por pozo es de: 500000.0',
 'El ingreso promedio del volumen obtenido es de: 309712.50000000006',
 'El beneficio promedio es de: -190287.4999999999')

In [20]:
metricas_beneficio(geo_data_2)

('El volumen para desarrollar un nuevo pozo sin pérdidas: 111.11111111111111',
 'El volumen promedio de las reservas previstas son: 95.00000000000004',
 'El gasto por pozo es de: 500000.0',
 'El ingreso promedio del volumen obtenido es de: 427500.00000000006',
 'El beneficio promedio es de: -72499.99999999993')

We can observe that the three regions present losses, since they have gains in negative values, in this case if we compare the values of the volume of reserves sufficient to develop a new well without losses with the average volume of each of the regions and we can observe that in all cases the volume necessary to avoid losses is greater, which indicates that the wells are not generating the volume necessary to avoid generating gains, therefore there are losses, however, region 2 is the one with less losses, which indicates that it is where the wells generate the most, so the region with the lowest losses is selected, since it is the one with the
The region with lower losses is selected, since the probability of finding wells with better yields is higher than the probability of the other two regions.

[Back to Contents](#back)

### Sample profit calculation

This time we will use a function similar to the one used to test the linear regression model, with the variant that in this one we will return the target and predictions that we used in the function to calculate the profit.

In [21]:
budget= 100000000
product_price= 4500
best_will = 200

In [22]:
target_predictions_0=pd.DataFrame(linear_df(geo_data_0)).transpose()
target_0= target_predictions_0['product']
predictions_0= target_predictions_0['predictions']

/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


El volúmen promedio de las reservas previstas son: 92.50000000000001
RECM del modelo de regresión lineal en el conjunto de validación:37.5794217150813


In [23]:
target_predictions_1=pd.DataFrame(linear_df(geo_data_1)).transpose()
target_1= target_predictions_1['product']
predictions_1= target_predictions_1['predictions']

/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


El volúmen promedio de las reservas previstas son: 68.82500000000002
RECM del modelo de regresión lineal en el conjunto de validación:0.893099286775617


In [24]:
target_predictions_2=pd.DataFrame(linear_df(geo_data_2)).transpose()
target_2= target_predictions_2['product']
predictions_2= target_predictions_2['predictions']

/tmp/ipykernel_27/898533389.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_valid['predictions']= predictions_valid


El volúmen promedio de las reservas previstas son: 95.00000000000004
RECM del modelo de regresión lineal en el conjunto de validación:40.02970873393434


In this case, a function is designed to determine the profit of the 200 best wells in each of the regions.

In [25]:
def profit(target,predictions):
    # Se acomodan los valores de las predicciones de orden descendiente 
    predictions_sorted= predictions.sort_values(ascending=False)
    # Se resumen y seleccionan los mejores 200 pozos con base a su producto
    selected_points = target[predictions_sorted.index[0:best_will]]
    # Se suman los valosres de la producción de los mejores 200 pozos
    product = selected_points.sum()
    # Se calcula la ganancia multiplicando la producción de barriles por el precio de $4'500 dólares
    revenue = product*product_price
    # Se establece el valor del presupuesto o del costo 200 pozos petroleros
    cost = budget

    
    return revenue - cost

In [26]:
profit(target_0,predictions_0)

33208260.43139851

In [27]:
profit(target_1,predictions_1)

24150866.966815114

In [28]:
profit(target_2,predictions_2)

27103499.635998324

Once the earnings of the 200 wells with the best barrel productions in each of the regions were calculated, we can see that region 0 is the one with the best 200 wells of the three regions, with a profit of 32.5 million dollars, while regions 1 and 2 have similar profits between 23 and 26 million dollars.

[Back to Contents](#back)

### Calculation of risks and profits for each region

A bootstrap function is now designed to obtain 1000 samples of 500 data, to select the 200 best wells from each of the samples, to calculate the average profit, the 95% confidence interval and the risk of loss.

In [29]:
# Variables such as botstrap size, sample size and random element are set to obtain 
state = np.random.RandomState(12345)
bootstrap_size= 1000
sample_size= 500

# Botstrap function
def bootstrap_reg(target,prediction):
    # profit_values variable is set
    profit_values=[]
    # Set the range with bootstrap size
    for i in range (bootstrap_size):
        # Target sample is set with the size of samples with random number generation with repetition 
        target_sample = target.sample(sample_size, replace=True, random_state=state)
        # Set the prediction sample
        predictions_sample = prediction[target_sample.index]
        
        profit_values.append(profit(target_sample, predictions_sample))
# The list of profit values in a series is established and the average profit is calculated
    profit_values = pd.Series(profit_values)
    mean_profit= profit_values.mean()
# Calculate the 95% confidence interval
    lower = profit_values.quantile(0.025)
    upper = profit_values.quantile(0.975)
# Calculate the downside risk by setting the number of wells with negative profit and divide 
    profit_negative= profit_values[profit_values<0].count()
    
    loss= 100* profit_negative/len(profit_values)

    return 'El beneficio promedio de la región es de: ' + str(mean_profit), 'Valor del intervalo más bajo: ' + str(lower), 'Valor del intervalo más alto: ' + str(upper), 'Valor del riesgo de pérdidas: ' + str(loss) + '%'
    


In [30]:
bootstrap_reg(target_0,predictions_0)

('El beneficio promedio de la región es de: 6007352.442611653',
 'Valor del intervalo más bajo: 129483.3113511391',
 'Valor del intervalo más alto: 12311636.057914983',
 'Valor del riesgo de pérdidas: 2.0%')

In [31]:
bootstrap_reg(target_1,predictions_1)

('El beneficio promedio de la región es de: 6639589.952601906',
 'Valor del intervalo más bajo: 2064763.6125177094',
 'Valor del intervalo más alto: 11911976.84748842',
 'Valor del riesgo de pérdidas: 0.1%')

In [32]:
bootstrap_reg(target_2,predictions_2)

('El beneficio promedio de la región es de: 5973810.479005231',
 'Valor del intervalo más bajo: 17349.29531177253',
 'Valor del intervalo más alto: 12462179.601652356',
 'Valor del riesgo de pérdidas: 2.5%')

We can observe that the average profit of region 1 is higher than the other two regions by approximately one million dollars, as well as we can observe that the risk of loss is twenty times lower in this zone than in the other two, region 1 only includes wells with profits in the lower interval value and its losses are less than 2.5%.

[Back to Contents](#back)

## Conclusions <a id='end'></a> 

In this project we initially performed an analysis of the data from each of the regions, then an analysis of the models that could possibly be used was made, 

Once the linear regression model was selected, functions were designed to calculate the profits of the best 200 wells in each region. Later, a bootstrap function was designed to obtain 1000 samples of 500 data, to select the best 200 wells of each sample, to calculate the average profit, the 95% confidence interval and the risk of losses by obtaining the number of lost wells in that sample.

In this work we found that :

1. being numerical data the regression models are those that we can use as the linear regression and forest, applying it for each of the regions, when analyzing it we found that it has a lower average error the linear regression model, which was selected to be able to use it in the analysis and design of functions of this project.
2. in the analysis of the profits of the best 200 wells of each of the regions, observing that the region that obtains more profit is region 0 with 33 million dollars, while regions 1 and 2 have similar profits between 24 and 27 million dollars.
3. the result of a bootstrap was that the average profit of 1000 samples from region 1 is a little more than 6 million dollars, due to the fact that the risk of loss is the lowest of 0.1, being 20 times less than that of region 0 and 25 times less than that of region 2.

Therefore, it is concluded that the best region for the development of oil wells is region 1.

[Back to Contents](#back)
